### PR TITLE
libraries:fatfs: Add wrapper for sd driver

### PR DESCRIPTION
**Pull request by mchindri**
Fatfs needs the implementation of physical device drivers.
This is our implementation of the Low level disk I/O module for FatFs.

Signed-off-by: Mihail Chindris <mihail.chindris@analog.com>

### CODE REVIEW

File: libraries/fatfs/adi_diskio.c [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+		return RES_NOTRDY;
+	case DEV_USB:
+		return RES_NOTRDY;
+	}
+	return RES_PARERR;
+}
+
+static int32_t init_power()
+{
+	if (adi_pwr_Init() != ADI_PWR_SUCCESS)
```
Comment (dbogdan) [inline libraries/fatfs/adi_diskio.c]: I would make this layer generic too. I guess adi_pwr_Init() is an ADuCM specific function.

**Comment by dbogdan** [comment]
This is not exactly an sd-card driver wrapper - it's our implementation of the Low level disk I/O module for FatFs.

**Comment by dbogdan** [comment]
Bu the way, this pull-request should not be merged before https://github.com/analogdevicesinc/no-OS/pull/372 does.

File: libraries/fatfs/adi_diskio.c [hunk 2]
```diff
@@ -0,0 +1,199 @@
+/*-----------------------------------------------------------------------*/
+/* Low level disk I/O module skeleton for FatFs     (C)ChaN, 2019        */
+/*-----------------------------------------------------------------------*/
+/* If a working storage control module is available, it should be        */
+/* attached to the FatFs via a glue function rather than modifying it.   */
+/* This is an example of glue functions to attach various exsisting      */
+/* storage control modules to the FatFs module with a defined API.       */
+/*-----------------------------------------------------------------------*/
```
Comment (mirceacaprioru) [inline libraries/fatfs/adi_diskio.c]: do we need an ADI copyright here ?

File: libraries/fatfs/adi_diskio.c [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+#include "sd.h"
+#include "error.h"
+#include <stdio.h>
+
+/* Definitions of physical drive number for each drive */
+#define DEV_SD		0	/* Example: Map MMC/SD card to physical drive 0 */
+#define DEV_RAM		1	/* Example: Map Ramdisk to physical drive 1 */
+#define DEV_USB		2	/* Example: Map USB MSD to physical drive 2 */
+
```
Comment (mirceacaprioru) [inline libraries/fatfs/adi_diskio.c]: nitpick: extra line

File: libraries/fatfs/adi_diskio.c [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+/*-----------------------------------------------------------------------*/
+/* Read Sector(s)                                                        */
+/*-----------------------------------------------------------------------*/
+
+DRESULT disk_read (
+	BYTE pdrv,		/* Physical drive nmuber to identify the drive */
+	BYTE *buff,		/* Data buffer to store read data */
+	LBA_t sector,		/* Start sector in LBA */
+	UINT count		/* Number of sectors to read */
+)
```
Comment (mirceacaprioru) [inline libraries/fatfs/adi_diskio.c]: this ) should go after the last member.

File: libraries/fatfs/adi_diskio.c [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	case DEV_SD :
+		return SD_disk_read(buff, sector, count);
+	case DEV_RAM :
+		return RES_NOTRDY;
+	case DEV_USB :
+		return RES_NOTRDY;
+	}
+	return RES_PARERR;
+}
+
```
Comment (mirceacaprioru) [inline libraries/fatfs/adi_diskio.c]: nitpick: extra lines

File: libraries/fatfs/adi_diskio.c [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+/*-----------------------------------------------------------------------*/
+
+DRESULT disk_read (
+	BYTE pdrv,		/* Physical drive nmuber to identify the drive */
+	BYTE *buff,		/* Data buffer to store read data */
+	LBA_t sector,		/* Start sector in LBA */
+	UINT count		/* Number of sectors to read */
+)
+{
+
```
Comment (mirceacaprioru) [inline libraries/fatfs/adi_diskio.c]: nitpick: extra line

File: libraries/fatfs/adi_diskio.c [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+/*-----------------------------------------------------------------------*/
+
+#if FF_FS_READONLY == 0
+
+DRESULT disk_write (
+	BYTE pdrv,			/* Physical drive nmuber to identify the drive */
+	const BYTE *buff,	/* Data to be written */
+	LBA_t sector,		/* Start sector in LBA */
+	UINT count			/* Number of sectors to write */
+)
```
Comment (mirceacaprioru) [inline libraries/fatfs/adi_diskio.c]: same as above

File: libraries/fatfs/adi_diskio.c [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+DSTATUS SD_disk_status(){
+	if (sd_init_var)
+		return 0;
+	return STA_NOINIT;
+}
+
+int32_t get_sd_card_init_param(struct sd_init_param *init_param);
+
+DSTATUS SD_disk_initialize(){
```
Comment (mirceacaprioru) [inline libraries/fatfs/adi_diskio.c]: } on the next line

File: libraries/fatfs/adi_diskio.c [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+		return RES_NOTRDY;
+	case DEV_USB :
+		return RES_NOTRDY;
+	}
+
+	return RES_PARERR;
+}
+
+#endif
+
```
Comment (mirceacaprioru) [inline libraries/fatfs/adi_diskio.c]: nitpick: extra line

File: libraries/fatfs/adi_diskio.c [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+	case DEV_RAM :
+		return STA_NODISK;
+	case DEV_USB :
+		return STA_NODISK;
+	default:
+		return STA_NODISK;
+	}
+	return STA_NOINIT;
+}
+
```
Comment (mirceacaprioru) [inline libraries/fatfs/adi_diskio.c]: nitpick: extra line

File: libraries/fatfs/adi_diskio.c [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+{
+	switch(pdrv) {
+	case DEV_SD:
+		switch (cmd){
+		case CTRL_SYNC: return RES_OK;
+		case GET_SECTOR_COUNT:
+			*(LBA_t *)buff = sd_desc->memory_size / DATA_BLOCK_LEN;
+			return RES_OK;
+		case GET_SECTOR_SIZE:
+			*(WORD *)buff = DATA_BLOCK_LEN;
```
Comment (mirceacaprioru) [inline libraries/fatfs/adi_diskio.c]: should this go in the next member GET_BLOCK_SIZE? 
also you could use a define here:
*(DWORD *)buff = 1u;

File: libraries/fatfs/adi_diskio.c [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+	if (sd_init_var == true)
+		return 0;
+	struct sd_init_param init_param;
+	if (SUCCESS != get_sd_card_init_param(&init_param))
+		return STA_NOINIT;
+	if (SUCCESS != sd_init(&sd_desc, &init_param))
+		return STA_NOINIT;
+	sd_init_var = true;
+	return 0;
```
Comment (mirceacaprioru) [inline libraries/fatfs/adi_diskio.c]: usally there is an extra line before this.

**Comment by mchindri** [comment]
V2 (For the last push)
 - Update commit description
 - Remove specific function and add function  `get_sd_card_init_param `that initialize the `init_param ` that must be implemented in the platform implementation

File: libraries/fatfs/adi_diskio.c (hunk 11 above)
Comment (mchindri) [reply libraries/fatfs/adi_diskio.c]: In the FatFs module the sector size is the same as the block size in the sd_card specification. That's why the naming is confusing.

**Comment by mchindri** [comment]
V3
 - add copyright header
 - remove surplus of '\n'
 - Put ) before the last parameter in functions

**Comment by mchindri** [comment]
V4
- Add comment in code for the confusion with sector and block size

**Comment by mchindri** [comment]
The possible solution for the init_param I see are:
 - Move the call to `sd_init `in main and make the `sd_desc` `extern` so it can be used for the read and write operations.
 - Add a function to `driver/platform/generic/platform_drivers.c` with a signature like: `int32_t configure_platform_param(void *param);` 
I think the first option will be more clean.

**Comment by buha** [comment]
I vote for 1st option with `sd_init` in `main`.

**Comment by mchindri** [comment]
v5:
 - Make the voted modification. Use sd_desc as extern

**Pull request merged**